File: src/84. Largest Rectangle in Histogram/84.go
func largestRectangleArea(heights []int) int {
	var stack []int
	stack = append(stack, -1)
	maxArea := 0
	for i := 0; i < len(heights); i++ {
		for len(stack) != 1 && heights[i] <= heights[stack[len(stack)-1]] {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1] // pop
			maxArea = maxVal(maxArea, heights[top]*(i-stack[len(stack)-1]-1))
		}
		stack = append(stack, i)
	}

	for len(stack) != 1 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		maxArea = maxVal(maxArea, heights[top]*(len(heights)-stack[len(stack)-1]-1))
	}

	return maxArea
}

func maxVal(a, b int) int {
	if a >= b {
		return a
	}
	return b
}

File: src/84. Largest Rectangle in Histogram/84.py
from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        heights = [0] + heights + [0]
        res = 0
        for i in range(len(heights)):
            while stack and heights[stack[-1]] > heights[i]:
                tmp = stack.pop()
                res = max(res, (i-stack[-1]-1)*heights[tmp])
            stack.append(i)
        return res
    def largestRectangleArea1(self, heights: List[int]) -> int:
        stack, res = [-1], 0
        for i in range(len(heights)):
            while stack[-1] != -1 and heights[stack[-1]] >= heights[i]:
                tmp = stack.pop()
                res = max(res, (i-stack[-1]-1)*heights[tmp])
            stack.append(i)
        return res


if __name__ == "__main__":
    heights = [2, 1, 5, 6, 2, 3]
    print(Solution().largestRectangleArea(heights))
    print(Solution().largestRectangleArea1(heights))
